### PR TITLE
Retries for HTTP requests

### PR DESCRIPTION
**Pull request by eujing**
This PR enables automatic retry and backoffs for HTTP requests made using requests, mainly targeting GETs.
This is to address intermittent connection issues observed during long-running operations such as polling, which would have resulted in OSErrors crashing MLOS and ending experiments prematurely.

The main patterns used are:
- https://docs.python-requests.org/en/latest/user/advanced/#example-automatic-retries
- https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#module-urllib3.util.retry

### CODE REVIEW

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py [hunk 1]
```diff
@@ -291,12 +309,16 @@ def _check_vm_operation_status(self, params: dict) -> Tuple[Status, dict]:
         if url is None:
             return Status.PENDING, {}
 
+        session = self._get_session(params)
         try:
-            response = requests.get(url, headers=self._get_headers(), timeout=self._request_timeout)
+            response = session.get(url, headers=self._get_headers(), timeout=self._request_timeout)
         except requests.exceptions.ReadTimeout:
             _LOG.warning("Request timed out: %s", url)
             # return Status.TIMED_OUT, {}
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/services/remote/azure/azure_services.py:317]:
```suggestion
```

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py [hunk 2]
```diff
@@ -201,6 +206,19 @@ def __init__(self,
...
 
+    def _get_session(self, params: dict) -> requests.Session:
+        total = params.get("requestTotalRetries", self._total_retries)
+        backoff_factor = params.get("requestBackoffFactor", self._backoff_factor)
+
+        session = requests.Session()
+        retry = Retry(total=total, backoff_factor=backoff_factor)
+        adapter = HTTPAdapter(max_retries=retry)
+
+        session.mount("http://", adapter)
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/services/remote/azure/azure_services.py:217]:
```suggestion
```

Don't think we should be supporting `http://` at all, right?

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py [hunk 3]
```diff
@@ -126,12 +184,46 @@ def test_wait_vm_operation_timeout(mock_requests: MagicMock,
...
+        "requestTotalRetries": total_retries,
+        "asyncResultsUrl": "https://DUMMY_ASYNC_URL",
+        "vmName": "test-vm",
+    }
+
+    # Simulate intermittent connection issues
+    mock_getconn.return_value.getresponse.side_effect = [
+        make_httplib_json_response(200, {"status": "InProgress"}),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
```
Comment (bpkroth) [inline mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:218]: repeat?

**Review by bpkroth — COMMENTED** [review]
some pytest errors it looks like, but mostly looks good

**Comment by bpkroth** [comment]
Might need to consider adding support for this to the SshServices as well.

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py [hunk 4]
```diff
@@ -291,12 +309,16 @@ def _check_vm_operation_status(self, params: dict) -> Tuple[Status, dict]:
         if url is None:
             return Status.PENDING, {}
 
+        session = self._get_session(params)
         try:
-            response = requests.get(url, headers=self._get_headers(), timeout=self._request_timeout)
+            response = session.get(url, headers=self._get_headers(), timeout=self._request_timeout)
```
Comment (motus) [inline mlos_bench/mlos_bench/services/remote/azure/azure_services.py:314]: we can probably move `headers=self._get_headers()` into the Session, too?

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py [hunk 5]
```diff
@@ -448,7 +470,16 @@ def _check_deployment(self, params: dict) -> Tuple[Status, dict]:
             deployment_name=config["deploymentName"],
         )
 
-        response = requests.get(url, headers=self._get_headers(), timeout=self._request_timeout)
+        session = self._get_session(params)
+        try:
+            response = session.get(url, headers=self._get_headers(), timeout=self._request_timeout)
```
Comment (motus) [inline mlos_bench/mlos_bench/services/remote/azure/azure_services.py:475]:
```suggestion
            response = session.get(url, timeout=self._request_timeout)
```
and move `headers=self._get_headers()` into the session

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py [hunk 6]
```diff
@@ -6,17 +6,75 @@
...
+    f = tempfile.TemporaryFile("wb+")
+    f.write(data)
+    f.seek(0)
+    f.status = status
+    f.msg = http.client.HTTPMessage()
+    f.length_remaining = len(data)
+    f.headers = {}
+    f.version = ""
+    f.reason = ""
+    f.isclosed = lambda: f.closed
```
[the remark this answers is not in the document]
Comment (eujing) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py]: This has been replaced with the actual `HTTPResponse` object, removing the need to manually set these attributes / methods.

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py [hunk 7]
```diff
@@ -6,17 +6,75 @@
...
+    mock_getconn.return_value.getresponse.side_effect = [
+        make_httplib_json_response(200, {"properties": {"provisioningState": "Running"}}),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
+        make_httplib_json_response(200, {"properties": {"provisioningState": "Running"}}),
+        make_httplib_json_response(200, {"properties": {"provisioningState": "Succeeded"}}),
+    ]
+
+    (status, _) = azure_vm_service.wait_host_deployment(params, is_setup=True)
+    assert status == operation_status
```
[the remark this answers is not in the document]
Comment (eujing) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:69]: Good point! I moved away from temp files to a in-memory `BytesIO` buffer, which should get cleaned up after tests exit

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py (hunk 3 above)
Comment (eujing) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:212]: Yeah I intended for this repeat to test cases of multiple retry attempts.
For example having >= 2 maximum retries should result in success, but having 1 maximum retry should still have a graceful failure.

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py (hunk 2 above)
Comment (eujing) [reply mlos_bench/mlos_bench/services/remote/azure/azure_services.py]: Removed `http://`, yes our API urls are all `https://`!

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py (hunk 4 above)
Comment (eujing) [reply mlos_bench/mlos_bench/services/remote/azure/azure_services.py]: Good point, moved the headers into Session too!

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py [hunk 8]
```diff
@@ -201,6 +206,20 @@ def __init__(self,
             with open(self._custom_data_file, 'r', encoding='utf-8') as custom_data_fh:
                 self._deploy_params["customData"] = custom_data_fh.read()
 
+    def _get_session(self, params: dict) -> requests.Session:
```
Comment (motus) [inline mlos_bench/mlos_bench/services/remote/azure/azure_services.py:209]: docstring?

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py [hunk 9]
```diff
@@ -201,6 +206,20 @@ def __init__(self,
...
+
+        session = requests.Session()
+        retry = Retry(total=total, backoff_factor=backoff_factor)
+        adapter = HTTPAdapter(max_retries=retry)
+
+        session.mount("https://", adapter)
+
+        session.headers.update(self._get_headers())
+
+        return session
```
Comment (motus) [inline mlos_bench/mlos_bench/services/remote/azure/azure_services.py:221]: I'd like more compact
```suggestion
        total_retries = params.get("requestTotalRetries", self._total_retries)
        backoff_factor = params.get("requestBackoffFactor", self._backoff_factor)
        session = requests.Session()
        session.mount("https://", HTTPAdapter(
            max_retries=Retry(total=total_retries, backoff_factor=backoff_factor)))
        session.headers.update(self._get_headers())
        return session
```
another nitpick, though

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py [hunk 10]
```diff
@@ -448,7 +470,16 @@ def _check_deployment(self, params: dict) -> Tuple[Status, dict]:
             deployment_name=config["deploymentName"],
         )
 
-        response = requests.get(url, headers=self._get_headers(), timeout=self._request_timeout)
+        session = self._get_session(params)
+        try:
+            response = session.get(url, timeout=self._request_timeout)
+        except requests.exceptions.ReadTimeout:
+            _LOG.warning("Request timed out: %s", url)
```
Comment (motus) [inline mlos_bench/mlos_bench/services/remote/azure/azure_services.py:477]:
```suggestion
            _LOG.warning("Request timed out after %.2f s: %s", self._request_timeout, url)
```

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py [hunk 11]
```diff
@@ -6,17 +6,69 @@
...
+    # Simulate intermittent connection issues
+    mock_getconn.return_value.getresponse.side_effect = [
+        make_httplib_json_response(200, {"properties": {"provisioningState": "Running"}}),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
+        make_httplib_json_response(200, {"properties": {"provisioningState": "Running"}}),
+        make_httplib_json_response(200, {"properties": {"provisioningState": "Succeeded"}}),
+    ]
+
+    (status, _) = azure_vm_service.wait_host_deployment(params, is_setup=True)
```
Comment (motus) [inline mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:68]: I'd move `params` closer to where we use it:
```suggestion
    (status, _) = azure_vm_service.wait_host_deployment(
        params={
            "pollInterval": 0.1,
            "requestTotalRetries": total_retries,
            "deploymentName": "TEST_DEPLOYMENT1",
            "subscription": "TEST_SUB1",
            "resourceGroup": "TEST_RG1",
        },
        is_setup=True
    )
```

**Review by motus — APPROVED** [review]
Left a few nitpicks, but looks good overall! Thank you!

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py (hunk 3 above)
Comment (bpkroth) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:212]: Maybe just add a comment explaining that.  Thanks!

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py [hunk 12]
```diff
@@ -6,17 +6,69 @@
...
+    """
+    Prepare a json response object for use with urllib3
+    """
+    data = json.dumps(json_data).encode("utf-8")
+    response = urllib3.HTTPResponse(
+        status=status,
+        body=BytesIO(data),
+        preload_content=False,
+    )
+    return response
```
[the remark this answers is not in the document]
Comment (bpkroth) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:34]: I'm fine with it the other way too.  I don't love chaining too many calls together in a single line and sometimes the explicit variable assignment can be very helpful for debugging (e.g., break points, pytest output, readability, etc.)

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py (hunk 3 above)
Comment (eujing) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:207]: Added the comment!

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (eujing) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py:34]: I left this as-is in favor of debug output in pytest!

File: mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py [hunk 13]
```diff
@@ -126,12 +178,46 @@ def test_wait_vm_operation_timeout(mock_requests: MagicMock,
...
+    # Simulate intermittent connection issues
+    mock_getconn.return_value.getresponse.side_effect = [
+        make_httplib_json_response(200, {"status": "InProgress"}),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
+        requests_ex.ConnectionError("Connection aborted", OSError(107, "Transport endpoint is not connected")),
+        make_httplib_json_response(200, {"status": "InProgress"}),
+        make_httplib_json_response(200, {"status": "Succeeded"}),
+    ]
+
+    (status, _) = azure_vm_service.wait_host_operation(params)
```
[the remark this answers is not in the document]
Comment (eujing) [reply mlos_bench/mlos_bench/tests/services/remote/azure/azure_services_test.py]: Agreed, moved it into the arguments

File: mlos_bench/mlos_bench/services/remote/azure/azure_services.py (hunk 8 above)
Comment (eujing) [reply mlos_bench/mlos_bench/services/remote/azure/azure_services.py:209]: Good catch! Added the docstring

**Review by motus — APPROVED** [review]
Looks great! let's merge it in

**Review by bpkroth — APPROVED** [review]
LGTM.  Thanks!

**Pull request merged**